### PR TITLE
Add MacBook touchbar support

### PR DESCRIPTION
**Pull request by paulroub**
Only enabled if (a) `useTouchbar` is enabled the configuration and (b) a touchbar is present

### CODE REVIEW

**Comment by paulroub** [comment]
I am really, really bad at `git rebase`.

File: lib/tasks.coffee [hunk 1]
```diff
@@ -73,6 +75,22 @@ module.exports =
       "tasks:convert-to-task": => @convertToTask()
       "tasks:set-timestamp": => @setTimestamp()
 
+    config = atom.config.get('tasks')
+    @useTouchbar = config.useTouchbar
+
+    _this = this
+
+    atom.config.onDidChange 'tasks.useTouchbar', ({newValue, oldValue}) ->
```
Comment (irrationalistic) [inline lib/tasks.coffee]: you can use `=>` to autobind context, so you shouldn't need the `_this` above!

File: lib/tasks.coffee [hunk 2]
```diff
@@ -73,6 +75,22 @@ module.exports =
...
+      _this.useTouchbar = newValue
+
+      if newValue != oldValue
+        _this.updateTouchbar()
+
+    @activeItemSub = atom.workspace.onDidChangeActivePaneItem =>
+      _this.subscribeToActiveTextEditor()
```
Comment (irrationalistic) [inline lib/tasks.coffee]: same here

File: lib/tasks.coffee [hunk 3]
```diff
@@ -133,8 +154,42 @@ module.exports =
...
 
 
+  ###*
+   * Update our active buttons based on the current line.
+  ###
+  updateTouchbar: ->
+    if @useTouchbar && tasks.checkIsTasks()
+      pt = @editor.getCursorBufferPosition()
+      config = atom.config.get('tasks')
+      linf = tasks.parseLine @editor, pt.row, config
```
Comment (irrationalistic) [inline lib/tasks.coffee]: linf?

File: lib/tasks.coffee [hunk 4]
```diff
@@ -292,6 +346,7 @@ module.exports =
     return if not editor
 
     selection = editor.getSelectedBufferRanges()
+    _this = this
```
Comment (irrationalistic) [inline lib/tasks.coffee]: can you use the autobinding to get rid of this?

File: lib/touchbar.coffee [hunk 5]
```diff
@@ -0,0 +1,57 @@
+tasks = require './tasksUtilities'
+
+{TouchBar} = require('remote')
+{TouchBarButton, TouchBarSpacer} = TouchBar
+
+module.exports =
+  update: (info, callback) ->
+    if not TouchBar
+      return
```
Comment (irrationalistic) [inline lib/touchbar.coffee]: you could do `return if not TouchBar`

File: lib/touchbar.coffee [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+      buttons.push @callbackButton("⇊", "Archive", callback)
+      buttons.push new TouchBarSpacer({size: 'small'})
+
+    touchBar = new TouchBar({items: buttons})
+
+    window.setTouchBar(touchBar)
+
+  callbackButton: (icon, command, callback, bgcolor) ->
+    opts = {
+      label: icon + " " + command,
```
Comment (irrationalistic) [inline lib/touchbar.coffee]: "#{icon} #{command}"

File: lib/views/task-status-view.coffee [hunk 7]
```diff
@@ -1,6 +1,7 @@
 tasks = require '../tasksUtilities'
 _ = require 'underscore'
 
+
```
Comment (irrationalistic) [inline lib/views/task-status-view.coffee]: probably leave this out :)

File: lib/views/task-status-view.coffee [hunk 8]
```diff
@@ -29,12 +30,14 @@ class TaskStatusView extends HTMLElement
...
   checkIsTasks: ->
-    if @editor?.getGrammar().name is 'Tasks'
+    if tasks.checkIsTasks()
       @style.display = ''
       return true
+
     @style.display = 'none'
     false
 
+
```
Comment (irrationalistic) [inline lib/views/task-status-view.coffee]: probably leave this out :)

File: lib/views/task-status-view.coffee [hunk 9]
```diff
@@ -63,5 +65,6 @@ class TaskStatusView extends HTMLElement
       total = '-' if isNaN total
       @textContent = "(#{completed}/#{total})"
 
+
```
Comment (irrationalistic) [inline lib/views/task-status-view.coffee]: same!

**Comment by irrationalistic** [comment]
This is super cool! I can't test it, but i trust you :)

One last thing: is it possible to get a screenshot of this to add to the readme? Just an example to show this cool feature off!

**Comment by paulroub** [comment]
Screen shot w/incomplete task:
screen-shot

Touchbar at that moment:
touchbar-shot

**Pull request merged**